python/ray/serve/_private/proxy_router.py

```diff
@@ -0,0 +1,189 @@
+                # or the next character in the target route is a '/'. This is
+                # to guard against the scenario where we have '/route' as a
+                # prefix and there's a request to '/routesuffix'. In this case,
+                # it should *not* be a match.
+                elif len(target_route) == len(route) or target_route[len(route)] == "/":
+                    matched = True
+
+                if matched:
+                    endpoint = self.route_info[route]
+                    return (
+                        route,
+                        self.handles[endpoint],
+                        endpoint.app,
+                        self.app_to_is_cross_language[endpoint.app],
+                    )
+
+        return None
+
+
+class EndpointRouter(ProxyRouter):
+    """Router that matches Endpoint on incoming routes."""
+
+    def __init__(self, get_handle: Callable):
+        # Function to get a handle given a name. Used to mock for testing.
+        self._get_handle = get_handle
+        # Contains a ServeHandle for each endpoint.
+        self.handles: Dict[EndpointTag, RayServeHandle] = dict()
+        # Endpoints info associated with endpoints.
+        self.endpoints: Dict[EndpointTag, EndpointInfo] = dict()
+
+    def update_routes(self, endpoints: Dict[EndpointTag, EndpointInfo]):
+        logger.info(
+            f"Got updated endpoints: {endpoints}.", extra={"log_to_stderr": False}
+        )
+        self.endpoints = endpoints
+        existing_handles = set(self.handles.keys())
+        for endpoint, info in endpoints.items():
+            if endpoint in self.handles:
+                existing_handles.remove(endpoint)
+            else:
+                self.handles[endpoint] = self._get_handle(
+                    endpoint.name, endpoint.app
+                ).options(
+                    # Streaming codepath isn't supported for Java.
+                    stream=(
+                        RAY_SERVE_ENABLE_EXPERIMENTAL_STREAMING
+                        and not info.app_is_cross_language
+                    ),
+                )
+
+        # Clean up any handles that are no longer used.
+        if len(existing_handles) > 0:
+            logger.info(
+                f"Deleting {len(existing_handles)} unused handles.",
+                extra={"log_to_stderr": False},
+            )
+        for endpoint in existing_handles:
+            del self.handles[endpoint]
+
+    def match_route(
+        self, target_route: str
+    ) -> Optional[Tuple[str, RayServeHandle, str, bool]]:
+        """Return the endpoint match among existing routes for the route.
+        Args:
+            target_route: endpoint to match against.
+        Returns:
+            (route, handle, app_name, is_cross_language) if found, else None.
+        """
+        for endpoint_tag, handle in self.handles.items():
+            if target_route == str(endpoint_tag):
```

Comment:
endpoint_tag is already a string?

Reply:
So with Cindy's recent change `endpoint_tag` is an object [`DeploymentID`](https://github.com/ray-project/ray/blob/f6930684b469cba16f34a016a0a06e1653de4019/python/ray/serve/_private/common.py#L33) We need to call `str()` to get the concatenated name for string comparison

Reply:
Oh, thanks for checking, I am fooled by the annotations, can you update it correspondingly?

Reply:
Which annotations are you talking about? 😅

Reply:
nvm, I was checking the old master...

Reply:
Thanks for checking!